### DISCUSSION TITLE
Is it possible to allocate heap memory?

### DISCUSSION THREAD

**Discussion by apppppppple** [aduros/wasm4#191] [Questions] [answered]
#138 makes it seem like you can allocate memory, but when I try to do it in zig with
```zig
const std = @import("std");
export fn update() void {
    const thing = std.heap.page_allocator.alloc(u32, 1) catch @panic("alloc failed!");
    std.heap.page_allocator.free(thing);
}
```
It seems to always panic.

**Comment by thedavesims** [comment, marked as answer]
I know nothing about zig but from a quick bit of research page_allocator looks like it requests an entire page, 64k. The total memory available to wasm4 is 64k but only ~58k bytes is available for program memory so page_allocator fails.

**Reply by apppppppple** [reply]
Thanks! Should just modifying the source to request less work? I'm not very familiar with pages.

**Reply by desttinghim** [reply]
I would recommend either using something like [zee_alloc](https://github.com/fengb/zee_alloc/) or the FixedBufferAllocator in the STD library.
Something like this:
```zig
const std = @import("std");
const heap: [100]u8 = undefined;
export fn update() void {
    const fba = std.heap.FixedBufferAllocator.init(&heap);
    alloc = fba.allocator();
    const thing = alloc.alloc(u32, 1) catch @panic("alloc failed!");
    alloc.free(thing);
}
```

**Reply by Mnohem** [reply]
This doesn't work for me. I get the error "Uncaught (in promise) LinkError: import object field 'memcpy' is not a Function" in the console when trying to use this, and the cart doesn't run. Is there some other boilerplate needed for wasm to use allocators?

**Reply by desttinghim** [reply]
There shouldn't be as far as I know, but it may be related to this issue here: https://github.com/ziglang/zig/issues/2910

If you have that workaround in your build file, see if turning that off fixes it.

**Reply by joshgoebel** [reply]
Do you perhaps need to add `memcpy` to `lib.export_symbol_names`?  I haven't played with any dynamic allocation stuff in Zig yet... I have in C and it "just worked" as expected.
